### PR TITLE
frontend: partly re-style the exchange page

### PR DESCRIPTION
**Pull request by shonsirsha**
as a part of the development of the exchange page, other than re-styling, some data (from the BE) also got re-structured to follow best practices in the FE.

[Preview] Video

https://user-images.githubusercontent.com/28676406/207245268-a5193140-c171-43a5-88a8-4afa865ef1d1.mov

This PR is a part of the development process and will be followed by one or several more PRs.

**What's still missing?**

1. Custom Select Component (Select Region) with Flags and Globe Icon
2. Payment Method Logo / Icon
3. The "Fast" and "Best Deal"  Badge
4. The "i" (information) button

**What's been done in this PR?**
1. Exchange Deals data re-structuring in the FE. This is done for several reasons, such as cleaning up the code, increasing modularity / re-usability of the components (decreasing "hardcoded-ness"), and ease of development. Will explain further in the comment section. 
2. Overall layout (spacings, alignments etc).

### CODE REVIEW

File: frontends/web/src/routes/buy/utils.tsx
Comment (shonsirsha) [inline frontends/web/src/routes/buy/utils.tsx:44]: These utils are created to change the data structure of the exchangeDeal (from BE). Now, the data that we use look like this:

The data is now restructured into this: `ExchangeProviderWithBestDeal` :


```js

type ExchangeProviderWithBestDeal = {
    providerName: string;
    methods: PaymentMethod[];
    supported?: boolean;
}

type PaymentMethod = ExchangeDeal & {
    isBestDeal: boolean;
}

//this ExchangeDeal already exists before
type ExchangeDeal = {
  fee: number;
  payment: 'card' | 'bank-transfer';
  isFast: boolean;
}
```

This way, it's now really easy to render all exchanges and its deals, including everything such as the "bestDeal" inside the object itself.

File: frontends/web/src/routes/buy/exchange.tsx
Comment (shonsirsha) [inline frontends/web/src/routes/buy/exchange.tsx:147]: I decided to re-structure the data using the util function I mentioned above rather than using this. There are several reasons, but it's difficult to manipulate the UI this way in react IMO. Other than that, we can't just append `details` with a JSX as it'd output a string (`[object Object]`).

Although this is good for the first version / to "prototype" the UI. 👍

File: frontends/web/src/routes/buy/exchange.tsx
Comment (shonsirsha) [inline frontends/web/src/routes/buy/exchange.tsx:174]: With the data already being re-structured and stored into the state `allExchangeProviders`, we can easily render all of them as the custom radio component like this. 

This means, when a new exchage provider is added in the future (which we'll most likely do), there will be a lot less code changes as we don't have to do

`{showMoonPay && moonpay element...}`
`{showPocket && pocket element...}`
`{showSomeOtherNewProvider1 && other provider1...}`
`{showSomeOtherNewProvider2 && other provider2...}`

etc.. 

as it already `.map` automatically.

**Comment by shonsirsha** [comment]
Lastly, this restructuring in the FE won't need to happen if the data from the BE already structured like this. IMO, it makes sense to actually return the data like below, which is what the FE doing atm:

```js
[
    {
        "providerName": "pocket",
        "methods": [
            {
                "fee": 0.015,
                "payment": "bank-transfer",
                "isFast": false,
                "isBestDeal": true
            }
        ],
        "supported": false
    },
    {
        "providerName": "moonpay",
        "methods": [
            {
                "fee": 0.019,
                "payment": "bank-transfer",
                "isFast": false,
                "isBestDeal": false
            },
            {
                "fee": 0.049,
                "payment": "card",
                "isFast": true,
                "isBestDeal": false
            }
        ],
        "supported": true
    }
]

```

This will reduce quite a lot of the code in the FE and as I wrote above, it's a lot easier to render things in the FE if we have the data structured "neatly" like this. Just my 2cents. 🙂

**Review by thisconnect — CHANGES_REQUESTED** [review]
Part I. reviewed.

will need more time for review part II.

File: frontends/web/src/components/forms/exchangeselectionradio.tsx
Comment (shonsirsha) [reply frontends/web/src/components/forms/exchangeselectionradio.tsx:62]: Yeah totally. This shouldve been "name" not "providerName". Good catch.

File: frontends/web/src/components/forms/exchangeselectionradio.module.css
Comment (shonsirsha) [reply frontends/web/src/components/forms/exchangeselectionradio.module.css:27]: Yes, as there's only one `label` in this component (radio).

Comment (shonsirsha) [reply frontends/web/src/components/forms/exchangeselectionradio.module.css:27]: I'll just assign a class for it.

File: frontends/web/src/routes/buy/exchange.tsx
Comment (shonsirsha) [reply frontends/web/src/routes/buy/exchange.tsx:139]: Hm.. I don't think putting a min-height here will stop it from jumping around as it'll try to centre it regardless (jumping coming from centreing this flex div). I will create a container for wrapping the content that changes the height, which are the radios and buttons and set the min-height here instead.

Comment (shonsirsha) [reply frontends/web/src/routes/buy/exchange.tsx:139]: I tried using View but can't seem to get the right settings to achieve the same result. Will tweak around a bit, can be done in the next PR if needed.

File: frontends/web/src/routes/buy/utils.tsx
Comment (shonsirsha) [reply frontends/web/src/routes/buy/utils.tsx:31]: I'm not using `const` here bc im using `let provider: keyof ExchangeDeals`. Reason for this is just to make TS happy (provider can't just be of "any" string, it has to be either pocket or moonpay).

Anyhow, sure. I'll use `Object.keys` & `.map` :)

File: frontends/web/src/components/forms/exchangeselectionradio.module.css
Comment (shonsirsha) [reply frontends/web/src/components/forms/exchangeselectionradio.module.css:46]: sure!

**Comment by shonsirsha** [comment]
Thanks for the review @thisconnect ! Changes made. PTAL. will wait for Part II review. 🙏

File: frontends/web/src/routes/buy/utils.tsx
Comment (thisconnect) [reply frontends/web/src/routes/buy/utils.tsx]: > I'm not using const here bc im using let provider: keyof ExchangeDeals.

ah, sorry missed that.

File: frontends/web/src/locales/en/app.json
Comment (thisconnect) [inline frontends/web/src/locales/en/app.json:316]: there is a linter check for translations, to fix just do: `make locize-fix` or alternatively (if you don't have locize installed globally) do: `npx locize-cli format frontends/web/src/locales --format json`

File: frontends/web/src/routes/buy/exchange.tsx
Comment (thisconnect) [reply frontends/web/src/routes/buy/exchange.tsx:141]: >  I don't think putting a min-height here will stop it from jumping around

if the min-height has the height of the content in largest possible form it should fix it.

Comment (shonsirsha) [reply frontends/web/src/routes/buy/exchange.tsx:141]: Hm, maybe Im misunderstanding this :)

here I set min-height to be 900px, which is plenty big

https://user-images.githubusercontent.com/28676406/207312673-5792208f-34bd-40bc-b410-45cbff38486f.mov
You can see it's still jumping around


I believe this is the reason, 


This is min-height 300:
Screen Shot 2022-12-13 at 13 16 50

This is min-height 1000:
Screen Shot 2022-12-13 at 13 06 18

To fix this, we can remove the current styling which involves flex shrink, flex grow, and justify content center. Instead, we can just use margin: auto, and setting the min height will work I believe.

Lmk what you think:)

Comment (thisconnect) [reply frontends/web/src/routes/buy/exchange.tsx:141]: looks good now 👍

**Comment by thisconnect** [comment]
I don't want to block this PR, but want to check findLowestFee function, see other comment.

There is one more issue, but I am not sure if it is part of this PR. On the pocket page it looks like there is 2 guides on top of each other.

Screen Shot 2022-12-13 at 13 41 09

File: frontends/web/src/routes/buy/exchange.tsx
Comment (shonsirsha) [reply frontends/web/src/routes/buy/exchange.tsx]: Made this change.

**Comment by shonsirsha** [comment]
> I don't want to block this PR, but want to check findLowestFee function, see other comment.
> 
> There is one more issue, but I am not sure if it is part of this PR. On the pocket page it looks like there is 2 guides on top of each other.
> 
> Screen Shot 2022-12-13 at 13 41 09

Yeah I noticed this too, but it already exists in the staging-pocket. I can create a fix in a separate PR to make sure we have clear context? :)

File: frontends/web/src/routes/buy/exchange.tsx
Comment (Beerosagos) [inline frontends/web/src/routes/buy/exchange.tsx:119]: I guess you should use `showPocket` (which depends on the account and the selected region) instead of `isPocketSupportedCoin` (which depends only on account). Otherwise, you won't filter correctly among regions. It seems to me that the current behavior shows both exchanges when at least one of theme is available. E.g. Pocket is available in Algeria, while it shouldn't (see https://widget.pocketbitcoin.com/api/availabilities for pocket availabilities)

**Comment by Beerosagos** [comment]
> Lastly, this restructuring in the FE won't need to happen if the data from the BE already structured like this. IMO, it makes sense to actually return the data from BE like below, which is what the FE doing atm:
> 
> ```js
> [
>     {
>         "providerName": "pocket",
>         "methods": [
>             {
>                 "fee": 0.015,
>                 "payment": "bank-transfer",
>                 "isFast": false,
>                 "isBestDeal": true
>             }
>         ],
>         "supported": false
>     },
>     {
>         "providerName": "moonpay",
>         "methods": [
>             {
>                 "fee": 0.019,
>                 "payment": "bank-transfer",
>                 "isFast": false,
>                 "isBestDeal": false
>             },
>             {
>                 "fee": 0.049,
>                 "payment": "card",
>                 "isFast": true,
>                 "isBestDeal": false
>             }
>         ],
>         "supported": true
>     }
> ]
> ```
> 
> This will reduce quite a lot of the code in the FE and as I wrote above, it's a lot easier to render things in the FE if we have the data structured "neatly" like this. This way, it's just a nicer DX and cleaner code. Just my 2cents. slightly_smiling_face Please lmk what you think! :)

Totally agree with you on this! If this is the best format to display data on FE, I can make a PR to refactor the endpoints and provide that. Lmk if you prefer to merge this and then have a PR that refactors the BE and simplifies everything or if it is better to merge the BE PR first and than rebase on that.
One important note: I avoided putting the `best deal` logic in the BE on purpose, since it depends on the region selections!

File: frontends/web/src/routes/buy/exchange.tsx
Comment (Beerosagos) [reply frontends/web/src/routes/buy/exchange.tsx:119]: Also, I think that the `best deal` logic should depend on the selected country. Otherwise, if the provider with the lowest fee is not available in a specific region we won't have a `best deal` option in that region (see the note in my other comment about that :) )

File: frontends/web/src/routes/buy/utils.tsx
Comment (Beerosagos) [inline frontends/web/src/routes/buy/utils.tsx:43]: The `best deal` tag makes sense only if there is more than one option. You could add a check on the number of available providers (after filtering them upon region)

Comment (shonsirsha) [reply frontends/web/src/routes/buy/utils.tsx]: Yup. makes sense. I'll add a check on the number of providers, but I'll keep `isBestDeal` property to exist regardless of the number of providers available (even if theres only 1 provider) to keep consistency. So `bestDeal` will always default to false when there's only one provider.

File: frontends/web/src/routes/buy/exchange.tsx
Comment (shonsirsha) [reply frontends/web/src/routes/buy/exchange.tsx]: Thanks for this! Will use `showPocket` and `showMoonPay` then.

File: frontends/web/src/routes/buy/utils.tsx
Comment (Beerosagos) [inline frontends/web/src/routes/buy/utils.tsx:66]: Filter and length check could be moved at beginning to skip the map if there are less than 2 providers? Just a nit, though

**Comment by Beerosagos** [comment]
> @Beerosagos Thanks for the comments! I've addressed them and made some changes. PTAL pray
> 
> > > Totally agree with you on this! If this is the best format to display data on FE, I can make a PR to refactor the endpoints and provide that.
> > > Lmk if you prefer to merge this and then have a PR that refactors the BE and simplifies everything or if it is better to merge the BE PR first and than rebase on that.
> 
> Yes, I think it'd be nice if we could have the BE to return [exactly like the above](https://github.com/digitalbitbox/bitbox-wallet-app/pull/1881#issuecomment-1347860782) without the `isBestDeal` (as you explained why).
> 
> Just a note, I refactored the util functions again. I think now it's more readable. But most importantly, if we refactor the BE to return exactly like the above (again without isBestDeal), we won't need to change much in this PR as I have isolated the "re-structuring" function and we can just remove this once the BE is ready +1
> 
> https://github.com/digitalbitbox/bitbox-wallet-app/blob/2be5e1fbf4ca4aadd8d69273241191539fdbb9bf/frontends/web/src/routes/buy/utils.tsx#L39-L56
> 
> Regarding which PR needs to be merged first, I think it'd be great if we can merge the BE PR first so this PR will then be 100% working solution already and there won't be a 2nd PR for the FE "correct" things due to change in the BE. Wdyt? slightly_smiling_face

Nice! I'll work on the refactoring and lyk when is ready! Thanks :)

**Comment by Beerosagos** [comment]
> Nice! I'll work on the refactoring and lyk when is ready! Thanks :)



> > Lastly, this restructuring in the FE won't need to happen if the data from the BE already structured like this. IMO, it makes sense to actually return the data from BE like below, which is what the FE doing atm:
> > ```js
> > [
> >     {
> >         "providerName": "pocket",
> >         "methods": [
> >             {
> >                 "fee": 0.015,
> >                 "payment": "bank-transfer",
> >                 "isFast": false,
> >                 "isBestDeal": true
> >             }
> >         ],
> >         "supported": false
> >     },
> >     {
> >         "providerName": "moonpay",
> >         "methods": [
> >             {
> >                 "fee": 0.019,
> >                 "payment": "bank-transfer",
> >                 "isFast": false,
> >                 "isBestDeal": false
> >             },
> >             {
> >                 "fee": 0.049,
> >                 "payment": "card",
> >                 "isFast": true,
> >                 "isBestDeal": false
> >             }
> >         ],
> >         "supported": true
> >     }
> > ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This will reduce quite a lot of the code in the FE and as I wrote above, it's a lot easier to render things in the FE if we have the data structured "neatly" like this. This way, it's just a nicer DX and cleaner code. Just my 2cents. slightly_smiling_face Please lmk what you think! :)
> 
> Totally agree with you on this! If this is the best format to display data on FE, I can make a PR to refactor the endpoints and provide that. Lmk if you prefer to merge this and then have a PR that refactors the BE and simplifies everything or if it is better to merge the BE PR first and than rebase on that. One important note: I avoided putting the `best deal` logic in the BE on purpose, since it depends on the region selections!

Endpoint refactor PR is ready: https://github.com/digitalbitbox/bitbox-wallet-app/pull/1892
I tried to stitch to @shonsirsha's request as much as possible.
A couple points:
1. Made a few changes in naming, to be coherent with backend code:
`provider` -> `exchange`
`method` -> `deal`
2. `supported` and `isBestDeal` depend on the selected region. This is why they are not filled in the backend, but I inserted them in the struct so that they can be used in FE without adding other types.

**Comment by shonsirsha** [comment]
Will go back to this PR once #1892 gets merged. Thanks.

**Comment by shonsirsha** [comment]
Ready for re-review! There's no change at all in the UI, but removed the part where the data gets re-structured as it's already structured nicely from the backend (due to PR #1892). 🙂 

Preview:

https://user-images.githubusercontent.com/28676406/208661150-d0be8535-c763-4926-aff8-afcded4d0f69.mov

**Review by Beerosagos — APPROVED** [review]
LGTM with just a couple nits :) Very nice!

File: frontends/web/src/api/exchanges.ts
Comment (Beerosagos) [inline frontends/web/src/api/exchanges.ts:47]: @thisconnect you preferred this field to be moved to a separate type, right?

Comment (thisconnect) [reply frontends/web/src/api/exchanges.ts:47]: yes, I think it makes sense that this type (actually all types in `api/*`) describes the data that is coming from the backend. Reading as is, it looks like the backend optionally sends supported boolean which it never sends.

So if `supported ` is added by the frontend, the type should be extended where this transformation happens.

File: frontends/web/src/routes/buy/utils.tsx
Comment (shonsirsha) [reply frontends/web/src/routes/buy/utils.tsx]: Did you mean if it's `false`? Anyhow, would you be able to suggest the code change? I don't think I understood you 😅 

What this map is doing is to mark `isBestDeal` is `true` if it's the lowest fee across all providers AND only when there's more than one exchanges. Otherwise it's `false`. Regardless, we want `isBestDeal` to always be there. Because before this .map(), isBestDeal is still not there yet.

**Comment by shonsirsha** [comment]
Changes made, PTAL :)

File: frontends/web/src/routes/buy/utils.tsx
Comment (Beerosagos) [reply frontends/web/src/routes/buy/utils.tsx]: > Did you mean if it's `false`? Anyhow, would you be able to suggest the code change? I don't think I understood you sweat_smile
> 
> What this map is doing is to mark `isBestDeal` is `true` if it's the lowest fee across all providers AND only when there's more than one exchanges. Otherwise it's `false`. Regardless, we want `isBestDeal` to always be there. Because before this .map(), isBestDeal is still not there yet.

Ahah right :D I wrote that before suggesting the type changes! Not valid anymore I guess :sweat_smile:

File: frontends/web/src/api/exchanges.ts
Comment (thisconnect) [reply frontends/web/src/api/exchanges.ts]: makes sense to me, latest change looks good to me, ping @Beerosagos

**Comment by Beerosagos** [comment]
> Changes made, PTAL :)

LGTM! Thanks for the patience :)

**Pull request merged**
